### PR TITLE
Add Arty A7 platform.

### PR DESCRIPTION
**Pull request by megng1**

### CODE REVIEW

File: nmigen_boards/arty_a7.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+        Resource("spi", 0,
+            Subsignal("clk",  Pins("F1",  dir="o")),
+            Subsignal("cs",   PinsN("C1", dir="o")),
+            Subsignal("mosi", Pins("H1",  dir="o")),
+            Subsignal("miso", Pins("G1",  dir="i")),
+            Attrs(IOSTANDARD="LVCMOS33")
+        ),
+
+        Resource("i2c", 0,
+            Subsignal("scl",        Pins("L18", dir="o")),
```
Comment (whitequark) [inline nmigen_boards/arty_a7.py]: This should be `dir="io"`, in case of a multimaster design.

File: nmigen_boards/arty_a7.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+        ),
+        Resource("eth", 0,
+            Subsignal("rst",     PinsN("C16", dir="o")),
+            Subsignal("mdio",    Pins("K13",  dir="io")),
+            Subsignal("mdc",     Pins("F16",  dir="o")),
+            Subsignal("rx_dv",   Pins("G16",  dir="i")),
+            Subsignal("rx_er",   Pins("C17",  dir="i")),
+            Subsignal("rx_data", Pins("D18 E17 E18 G17", dir="i")),
+            Subsignal("tx_en",   Pins("H15",  dir="o")),
+            Subsignal("tx_data", Pins("H14 J14 J13 H17", dir="o")),
```
Comment (whitequark) [inline nmigen_boards/arty_a7.py:117]: I'm a bit confused--is this RGMII? Doesn't it have `dv` and `er` combined into one signal?

Comment (megng1) [reply nmigen_boards/arty_a7.py:117]: According to the [manual](https://reference.digilentinc.com/reference/programmable-logic/arty-a7/reference-manual#ethernet_phy), the PHY has a MII interface.

Comment (whitequark) [reply nmigen_boards/arty_a7.py:117]: Ah sorry, I forgot that MII has half the pins of GMII.

File: nmigen_boards/arty_a7.py [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+            Subsignal("reset", PinsN("K6", dir="o"), Attrs(IOSTANDARD="SSTL135")),
+            Attrs(SLEW="FAST"),
+        ),
+
+        Resource("eth_ref_clk", 0, Pins("G18", dir="o"), Attrs(IOSTANDARD="LVCMOS33")),
+        Resource("eth_clocks", 0,
+            Subsignal("tx", Pins("H16", dir="i")),
+            Subsignal("rx", Pins("F15", dir="i")),
+            Attrs(IOSTANDARD="LVCMOS33")
+        ),
```
Comment (whitequark) [inline nmigen_boards/arty_a7.py]: I think these clocks should be a part of the `eth_mii` resource. (That's what I've done for Versa ECP5).

File: nmigen_boards/arty_a7.py [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+        Resource("eth", 0,
```
Comment (whitequark) [inline nmigen_boards/arty_a7.py]: This should be called `eth_mii`.

File: nmigen_boards/arty_a7.py [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+                DiffPairs("U9", "V9", dir="o"),
+                Attrs(IOSTANDARD="DIFF_SSTL135")
+            ),
+            Subsignal("cke",   Pins("N5",  dir="o"), Attrs(IOSTANDARD="SSTL135")),
+            Subsignal("odt",   Pins("R5",  dir="o"), Attrs(IOSTANDARD="SSTL135")),
+            Subsignal("reset", PinsN("K6", dir="o"), Attrs(IOSTANDARD="SSTL135")),
+            Attrs(SLEW="FAST"),
+        ),
+
+        Resource("eth_ref_clk", 0, Pins("G18", dir="o"), Attrs(IOSTANDARD="LVCMOS33")),
```
Comment (whitequark) [inline nmigen_boards/arty_a7.py]: Does this clock have a fixed frequency? Or does the reference clock vary too in MII?

File: nmigen_boards/arty_a7.py [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+            Subsignal("sda_pullup", Pins("A13", dir="o")),
+            Attrs(IOSTANDARD="LVCMOS33")
+        ),
+
+        *SPIFlashResources(0,
+            cs="L13", clk="L16", mosi="K17", miso="K18", wp="L14", hold="M14",
+            attrs=Attrs(IOSTANDARD="LVCMOS33")
+        ),
+
+        Resource("ddram", 0,
```
Comment (whitequark) [inline nmigen_boards/arty_a7.py]: I think this should be called `ddr2`, `ddr3`, etc depending on actual version.

File: nmigen_boards/arty_a7.py [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+
+        Resource("serial", 0,
+            Subsignal("tx", Pins("D10", dir="o")),
+            Subsignal("rx", Pins("A9",  dir="i")),
+            Attrs(IOSTANDARD="LVCMOS33")
+        ),
+
+        Resource("cpu_reset", 0, Pins("C2", dir="o"), Attrs(IOSTANDARD="LVCMOS33")),
+
+        Resource("spi", 0,
```
Comment (whitequark) [inline nmigen_boards/arty_a7.py]: Would you be willing to add `SPIResource` like `SPIFlashResources`?

File: nmigen_boards/arty_a7.py [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+            Subsignal("rst",     PinsN("C16", dir="o")),
+            Subsignal("mdio",    Pins("K13",  dir="io")),
+            Subsignal("mdc",     Pins("F16",  dir="o")),
+            Subsignal("rx_dv",   Pins("G16",  dir="i")),
+            Subsignal("rx_er",   Pins("C17",  dir="i")),
+            Subsignal("rx_data", Pins("D18 E17 E18 G17", dir="i")),
+            Subsignal("tx_en",   Pins("H15",  dir="o")),
+            Subsignal("tx_data", Pins("H14 J14 J13 H17", dir="o")),
+            Subsignal("col",     Pins("D17",  dir="i")),
+            Subsignal("crs",     Pins("G14",  dir="i")),
```
Comment (whitequark) [inline nmigen_boards/arty_a7.py]: These are receiver signals logically, so I'd say these should be called `rx_col` and `rx_crs`.

**Review by whitequark — CHANGES_REQUESTED** [review]

File: nmigen_boards/arty_a7.py (hunk 5 above)
Comment (megng1) [reply nmigen_boards/arty_a7.py]: Yes, its frequency must be 25MHz for MII, and 50 MHz for RMII.

File: nmigen_boards/arty_a7.py (hunk 7 above)
Comment (megng1) [reply nmigen_boards/arty_a7.py]: Sure, I will do it in a later commit.

File: nmigen_boards/arty_a7.py (hunk 8 above)
Comment (whitequark) [reply nmigen_boards/arty_a7.py]: (I was wrong: in MII, COL and CRS are asynchronous.)

File: nmigen_boards/arty_a7.py [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+            Attrs(IOSTANDARD="LVCMOS33")
+        ),
+        Resource("eth_rmii", 0,
+            Subsignal("rst",       PinsN("C16", dir="o")),
+            Subsignal("mdio",      Pins("K13", dir="io")),
+            Subsignal("mdc",       Pins("F16", dir="o")),
+            Subsignal("tx_en",     Pins("H15", dir="o")),
+            Subsignal("tx_data",   Pins("H14 J14", dir="o")),
+            Subsignal("rx_crs_dv", Pins("G14", dir="i")),
+            Subsignal("rx_dv",     Pins("G16", dir="i"), Attrs(PULLUP="TRUE")),
```
Comment (whitequark) [inline nmigen_boards/arty_a7.py]: OK, I believe there is only one remaining review item. Could you explain what is `rx_dv` here? AFAIK, in RMII there is only `rx_crs_dv`.

Comment (megng1) [reply nmigen_boards/arty_a7.py]: The PHY (TI DP83848J) can operate in RMII mode by strapping `rx_dv` high during reset.
`rx_dv` can also be used in RMII mode as a simpler way to access DV, without having to extract it from `rx_crs_dv`.

This is according to the datasheet, I actually haven't tested RMII on the Arty.

**Comment by whitequark** [comment]
Thanks!

**Pull request merged**